File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/Hotel.py
class Hotel:
    def __init__(self, name, rooms):
        self.name = name
        self.availableRooms = dict(rooms)  # shallow copy
        self.bookedRooms = {}

    def bookRoom(self, roomType, roomNumber, name):
        if roomType not in self.availableRooms:
            return "False"
        available = self.availableRooms[roomType]
        if roomNumber <= available:
            self.bookedRooms.setdefault(roomType, {})
            self.bookedRooms[roomType][name] = roomNumber
            self.availableRooms[roomType] = available - roomNumber
            return "Success!"
        else:
            return "False"

    def checkIn(self, roomType, roomNumber, name):
        if roomType not in self.bookedRooms or name not in self.bookedRooms[roomType]:
            return False
        booked = self.bookedRooms[roomType][name]
        if roomNumber > booked:
            return False
        elif roomNumber == booked:
            del self.bookedRooms[roomType][name]
        else:
            self.bookedRooms[roomType][name] = booked - roomNumber
        return True

    def checkOut(self, roomType, roomNumber):
        self.availableRooms[roomType] = self.availableRooms.get(roomType, 0) + roomNumber

    def getAvailableRooms(self, roomType):
        return self.availableRooms.get(roomType, 0)


if __name__ == "__main__":
    rooms = {"single": 3, "double": 2}
    hotel = Hotel("Test Hotel", rooms)

    print(hotel.bookRoom("single", 2, "guest 1"))
    print(hotel.bookRoom("triple", 2, "guest 1"))
    print(hotel.bookRoom("single", 2, "guest 2"))
    print(hotel.bookRoom("single", 1, "guest 2"))
    print(hotel.bookRoom("single", 3, "guest 1"))
    print(hotel.bookRoom("single", 100, "guest 1"))

    hotel.checkIn("single", 1, "guest 1")
    print(hotel.bookedRooms)
    print(hotel.checkIn("single", 3, "guest 1"))
    print(hotel.checkIn("double", 1, "guest 1"))
    hotel.checkIn("double", 1, "guest 2")
    print(hotel.bookedRooms)

    hotel.checkOut("single", 1)
    print(hotel.availableRooms)
    hotel.checkOut("triple", 2)
    print(hotel.availableRooms)

    print(hotel.getAvailableRooms("single"))

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_cpp/code_output/Hotel.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <sstream>

std::string mapToString(const std::unordered_map<std::string, int>& m) {
    std::ostringstream oss;
    oss << "{";
    bool first = true;
    for (const auto& [k, v] : m) {
        if (!first) oss << ", ";
        first = false;
        oss << k << "=" << v;
    }
    oss << "}";
    return oss.str();
}

std::string nestedMapToString(const std::unordered_map<std::string, std::unordered_map<std::string, int>>& m) {
    std::ostringstream oss;
    oss << "{";
    bool first = true;
    for (const auto& [k, v] : m) {
        if (!first) oss << ", ";
        first = false;
        oss << k << "=" << mapToString(v);
    }
    oss << "}";
    return oss.str();
}

class Hotel {
public:
    std::string name;
    std::unordered_map<std::string, int> availableRooms;
    std::unordered_map<std::string, std::unordered_map<std::string, int>> bookedRooms;

    Hotel(const std::string& name, const std::unordered_map<std::string, int>& rooms)
        : name(name), availableRooms(rooms), bookedRooms() {}

    std::string bookRoom(const std::string& roomType, int roomNumber, const std::string& guestName) {
        if (availableRooms.find(roomType) == availableRooms.end()) {
            return "False";
        }
        int available = availableRooms.at(roomType);
        if (roomNumber <= available) {
            if (bookedRooms.find(roomType) == bookedRooms.end()) {
                bookedRooms[roomType] = std::unordered_map<std::string, int>();
            }
            bookedRooms[roomType][guestName] = roomNumber;
            availableRooms[roomType] = available - roomNumber;
            return "Success!";
        } else {
            return "False";
        }
    }

    bool checkIn(const std::string& roomType, int roomNumber, const std::string& guestName) {
        auto outerIt = bookedRooms.find(roomType);
        if (outerIt == bookedRooms.end()) return false;
        auto innerIt = outerIt->second.find(guestName);
        if (innerIt == outerIt->second.end()) return false;

        int booked = innerIt->second;
        if (roomNumber > booked) {
            return false;
        } else if (roomNumber == booked) {
            outerIt->second.erase(guestName);
        } else {
            outerIt->second[guestName] = booked - roomNumber;
        }
        return true;
    }

    void checkOut(const std::string& roomType, int roomNumber) {
        auto it = availableRooms.find(roomType);
        if (it == availableRooms.end()) {
            availableRooms[roomType] = roomNumber;
        } else {
            it->second += roomNumber;
        }
    }

    int getAvailableRooms(const std::string& roomType) {
        auto it = availableRooms.find(roomType);
        if (it == availableRooms.end()) {
            return 0;
        }
        return it->second;
    }
};

int main() {
    std::unordered_map<std::string, int> rooms;
    rooms["single"] = 3;
    rooms["double"] = 2;
    Hotel hotel("Test Hotel", rooms);

    std::cout << hotel.bookRoom("single", 2, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("triple", 2, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("single", 2, "guest 2") << std::endl;
    std::cout << hotel.bookRoom("single", 1, "guest 2") << std::endl;
    std::cout << hotel.bookRoom("single", 3, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("single", 100, "guest 1") << std::endl;

    hotel.checkIn("single", 1, "guest 1");
    std::cout << nestedMapToString(hotel.bookedRooms) << std::endl;
    std::cout << std::boolalpha << hotel.checkIn("single", 3, "guest 1") << std::endl;
    std::cout << std::boolalpha << hotel.checkIn("double", 1, "guest 1") << std::endl;
    hotel.checkIn("double", 1, "guest 2");
    std::cout << nestedMapToString(hotel.bookedRooms) << std::endl;

    hotel.checkOut("single", 1);
    std::cout << mapToString(hotel.availableRooms) << std::endl;
    hotel.checkOut("triple", 2);
    std::cout << mapToString(hotel.availableRooms) << std::endl;

    std::cout << hotel.getAvailableRooms("single") << std::endl;

    return 0;
}